pkg/controller/datavolume/snapshot-clone-controller.go

```diff
@@ -0,0 +1,486 @@
+	// converting volume mode is possible but has security implications
+
+	// Sizes validation
+	restoreSize := snapshot.Status.RestoreSize
+	if restoreSize == nil {
+		return true, fmt.Errorf("snapshot has no RestoreSize")
+	}
+	targetRequest, hasTargetRequest := pvcSpec.Resources.Requests[corev1.ResourceStorage]
+	allowExpansion := targetStorageClass.AllowVolumeExpansion != nil && *targetStorageClass.AllowVolumeExpansion
+	if hasTargetRequest {
+		// otherwise will just use restoreSize
+		if restoreSize.Cmp(targetRequest) < 0 && !allowExpansion {
+			r.log.V(3).Info("Can't expand restored PVC because SC does not allow expansion, need to fall back to host assisted")
+			return true, nil
+		}
+	}
+
+	if !isCrossNamespaceClone(datavolume) || *bindingMode == storagev1.VolumeBindingImmediate {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (r *SnapshotCloneReconciler) reconcileRestoreSnapshot(log logr.Logger,
+	datavolume *cdiv1.DataVolume,
+	snapshot *snapshotv1.VolumeSnapshot,
+	pvcSpec *corev1.PersistentVolumeClaimSpec,
+	transferName string,
+	syncRes *dataVolumeCloneSyncResult) (reconcile.Result, error) {
+
+	pvcName := datavolume.Name
+
+	if isCrossNamespaceClone(datavolume) {
+		pvcName = transferName
+		result, err := r.doCrossNamespaceClone(log, syncRes, pvcName, datavolume.Spec.Source.Snapshot.Namespace, false, SmartClone)
+		if result != nil {
+			return *result, err
+		}
+	}
+
+	if datavolume.Status.Phase == cdiv1.NamespaceTransferInProgress {
+		return reconcile.Result{}, nil
+	}
+
+	newPvc, err := r.makePvcFromSnapshot(pvcName, datavolume, snapshot, pvcSpec)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	currentRestoreFromSnapshotPvc := &corev1.PersistentVolumeClaim{}
+	if err := r.client.Get(context.TODO(), client.ObjectKeyFromObject(newPvc), currentRestoreFromSnapshotPvc); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+		if err := r.client.Create(context.TODO(), newPvc); err != nil {
+			if cc.ErrQuotaExceeded(err) {
+				r.syncDataVolumeStatusPhaseWithEvent(syncRes, cdiv1.Pending, nil,
+					Event{
+						eventType: corev1.EventTypeWarning,
+						reason:    cc.ErrExceededQuota,
+						message:   err.Error(),
+					})
+			}
+			return reconcile.Result{}, err
+		}
+	} else {
+		if currentRestoreFromSnapshotPvc.Status.Phase == corev1.ClaimBound {
+			if err := r.setCloneOfOnPvc(currentRestoreFromSnapshotPvc); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	}
+
+	return reconcile.Result{}, r.syncCloneStatusPhase(syncRes, cdiv1.CloneFromSnapshotSourceInProgress, nil)
+}
+
+func (r *SnapshotCloneReconciler) createTempHostAssistedSourcePvc(dv *cdiv1.DataVolume, snapshot *snapshotv1.VolumeSnapshot, targetPvcSpec *corev1.PersistentVolumeClaimSpec, syncRes *dataVolumeCloneSyncResult) error {
+	tempHostAssistedSourcePvc, err := r.makePvcFromSnapshot(snapshot.Name, dv, snapshot, targetPvcSpec)
```

Comment:
it `snapshot.Name` is not a good name for the temp pvc.  what about concurrent clones?  The create will fail for the second.

It would of course be nice to have a single shared temp PVC but I don't think that was your intention.  And implementing that scheme would add complexity.  Maybe a subsequent pr?

Reply:
You're right, it was not my intention, I am cleaning up the "shared" PVC once the corresponding DV is done.
BTW - the creation won't actually fail. If the PVC exists, we will just attempt to use it.

Just to make sure I understand, do you mean we should switch to unique PVC per fallback dumb clone for this PR,
and follow up later with attempts to clone from shared?

BTW' - I played around with concurrent host-assisted clones (both PVC & snapshot source)
Seems we have a problem outside of this PR, the source pod gets restarted multiple times before completing
```bash
I0122 14:58:03.462143      10 prometheus.go:72] 25.53
F0122 14:58:03.536422      10 clone-source.go:41] Subprocess did not execute successfully, result is: '\x02'
disk.img
/usr/bin/tar: disk.img: Read error at byte 1119744, while reading 512 bytes: Permission denied
/usr/bin/tar: Exiting with failure status due to previous errors

NAMESPACE   NAME                  PHASE             PROGRESS   RESTARTS   AGE
default     cloned-datavolume-1   CloneInProgress   24.66%     3          2m13s
```

Reply:
> Just to make sure I understand, do you mean we should switch to unique PVC per fallback dumb clone for this PR,
and follow up later with attempts to clone from shared?

yup

Reply:
So yeah shared source seems to work, but it may be ugly that we try deleting it, should we protect it with `GetPodsUsingPVCs`?

Reply:
May work but I'm sure is racy.  Unique source pvcs is the way to go for now